File: src/boards/rectangular.ts
import {BaseBoard, hasInterWall, isEnabled, setInterWallValue} from "../base.js";
import {PartialExcept} from "../types";
import {keys} from "ts-transformer-keys";

/*--------------
 * Types
 *-------------- */

export enum Direction {
  TOP = 0b0001,
  RIGHT = 0b0010,
  BOTTOM = 0b0100,
  LEFT = 0b1000,
}

export interface Size {
  height: number;
  width: number;
}

export interface Position {
  x: number;
  y: number;
}

export interface RectangularBoard extends BaseBoard {
  size: Size;
}

export const _size_params = keys<Size>();

/*-------------------------
 * Constructor Functions
 *------------------------- */

/**
 * Returns a new rectangularBoard for the given size
 */
export function newBoard(size: Size): RectangularBoard {
  return {
    cells: new Uint8Array(size.width * size.height),
    size,
  }
}

/**
 * Casts base board to RectangularBoard
 */
export function newFromBaseBoard({cells}: BaseBoard, size: Size): RectangularBoard {
  return {cells: cells, size}
}

/*-------------------------
 * Position Functions
 *------------------------- */

/**
 * Linear index from position
 */
export function toIndex(position: Position, {size}: PartialExcept<RectangularBoard, 'size'>) {
  return position.y * size.width + position.x;
}

/**
 * Position from linear index
 */
export function toPosition(index: number, {size}: PartialExcept<RectangularBoard, 'size'>) {
  return {
    x: index % size.width,
    y: Math.floor(index / size.width),
  };
}

/*-------------------------
 * Cell value Functions
 *------------------------- */

/**
 * get cell at given position
 */
export function getCellByPosition(position: Position, {cells, size}: RectangularBoard): number {
  return cells[toIndex(position, {size})];
}

/**
 * set cell at given position
 */
export function setCellByPosition(position: Position, value: number, {cells, size}: RectangularBoard) {
  return cells[toIndex(position, {size})] = value;
}

/*-------------------------
 * Direction Utils
 *------------------------- */

/**
 * return a opposing direction
 *
 * getOpposingDirection(Direction.LEFT) -> Direction.RIGHT
 */
export function getOpposingDirection(direction: Direction): Direction {
  return ((direction << 2) | (direction >> 2)) & 0b1111;
}

/**
 * Get relative direction between two positions
 */
export function getRelativeDirection(
  index1: number, index2: number,
  {size}: PartialExcept<RectangularBoard, 'size'>
): Direction {
  const pos1 = toPosition(index1, {size});
  const pos2 = toPosition(index2, {size});

  if (pos1.y > pos2.y) return Direction.TOP;
  if (pos1.x < pos2.x) return Direction.RIGHT;
  if (pos1.y < pos2.y) return Direction.BOTTOM;
  if (pos1.x > pos2.x) return Direction.LEFT;

  throw `'${pos1}' and '${pos2}' are not neighbours`;
}

/*-------------------------
 * Cell Neighbourhood Utils
 *------------------------- */

/**
 * Returns a new position in direction relative to the given position
 */
export function getRelativePosition({x, y}: Position, direction: Direction) {
  let newPosition = {x, y};
  if (direction === Direction.TOP) newPosition.y--;
  if (direction === Direction.RIGHT) newPosition.x++;
  if (direction === Direction.BOTTOM) newPosition.y++;
  if (direction === Direction.LEFT) newPosition.x--;
  return newPosition;
}

/**
 * Get neighbour cells of the given position
 */
export function getNeighbours(index: number, {cells, size}: RectangularBoard): number[] {
  let neighboursCells = [];

  // TOP
  if (index >= size.width) { neighboursCells.push(index - size.width); }
  // RIGHT
  if ((index + 1) % size.width != 0) { neighboursCells.push(index + 1); }
  // BOTTOM
  if (index < cells.length - size.width) { neighboursCells.push(index + size.width); }
  // LEFT
  if (index % size.width != 0) { neighboursCells.push(index - 1); }

  neighboursCells = neighboursCells.filter((i) => isEnabled(cells[i]));

  return neighboursCells;
}

/**
 * get allowed directions from a given position
 *
 * if visitableOnly is false then it only check nif neighbour is enabled or not
 */
export function getAllowedDirection({x, y}: Position, {cells, size}: RectangularBoard, visitableOnly = true) {
  let directions = [];

  if (y > 0) directions.push(Direction.TOP);
  if (x < size.width - 1) directions.push(Direction.RIGHT);
  if (y < size.height - 1) directions.push(Direction.BOTTOM);
  if (x > 0) directions.push(Direction.LEFT);

  directions.filter((dir) => {
    const newPos = getRelativePosition({x, y}, dir);
    const index1 = toIndex(newPos, {size});
    const index2 = toIndex({x, y}, {size});
    const cell = cells[index1];
    if (visitableOnly && hasInterWall<RectangularBoard, Direction>(
      index1, index2, {cells: cells, size},
      getRelativeDirection, getOpposingDirection
    )) {
      return false;
    }
    return isEnabled(cell);
  });

  return directions;
}

/**
 * return array of rows of cells
 */
export function getRows({cells, size}: RectangularBoard): number[][] {
  return cells
    // map cell to its index
    .map((_, i) => i)
    // map indexes into rows based on board width
    .reduce((acc, item, index) => {
      if (index % size.width === 0) {
        acc.push([]);
      }

      acc[acc.length - 1].push(item);
      return acc;
    }, [])
    // don't allow any disabled cell
    .map((row) => row.filter((c) => isEnabled(cells[c])))
    // ignore empty rows
    .filter((row) => row.length);
}

/*-------------------------
 * Cell Wall Utils
 *------------------------- */

/**
 * Remove wall between the given two cell Indexes
 */
export function removeInterWall(index1: number, index2: number, board: RectangularBoard): RectangularBoard {
  return setInterWallValue<RectangularBoard, Direction>(
    index1, index2, board,
    getOpposingDirection,
    getRelativeDirection,
    (cell, dir) => cell | dir
  )
}

/**
 * Set wall between the given two cell positions
 */
export function setInterWall(index1: number, index2: number, board: RectangularBoard): RectangularBoard {
  return setInterWallValue<RectangularBoard, Direction>(
    index1, index2, board,
    getOpposingDirection,
    getRelativeDirection,
    (cell, dir) => cell & ~dir
  )
}

File: python/boards/rectangular.py
from ..base import BaseBoard

class Direction:
    TOP = 0b0001
    RIGHT = 0b0010
    BOTTOM = 0b0100
    LEFT = 0b1000

# class Position:
#     def __int__(self,x,y):
#         self.x = x
#         self.y = y
    
#     @classmethod
#     def from_index(cls,index):
        

class RectangularBoard(BaseBoard):

    def __init__(self,width,height):
        self.width = width
        self.height = height
        
        super().__init__(self.width*self.height,"rectangular")

    def to_index(self,x,y):
        return y*self.width+x
    
    def to_position(self,index):
        return index%self.width,index//self.width
    
    def get_cell_by_position(self,x,y):
        return self.cells[self.to_index(x,y)]

    def set_cell_by_position(self,x,y,value):
        self.cells[self.to_index(x,y)] = value
    
    def get_opposing_direction(self,direction):
        return ((direction << 2) | (direction >> 2)) & 0b1111

    def get_relative_direction(self,index1,index2):
        pos1x,pos1y = self.to_position(index1)
        pos2x,pos2y = self.to_position(index2)

        if (pos1y > pos2y):
            return Direction.TOP
        if (pos1x < pos2x):
            return Direction.RIGHT
        if (pos1y < pos2y):
            return Direction.BOTTOM
        if (pos1x > pos2x):
            return Direction.LEFT

        raise Exception(f"({pos1x}, {pos1y}) and ({pos2x}, {pos2y})  are not neighbours")

    def get_relative_position(self,x,y,direction):
        if (direction == Direction.TOP):
            y-=1
        if (direction == Direction.RIGHT):
            x+=1
        if (direction == Direction.BOTTOM):
            y+=1
        if (direction == Direction.LEFT):
            x-=1
        return x,y
    
    def get_neighbours(self,index):

        neighbours_cells = []
        # TOP
        if (index >= self.width):
            neighbours_cells.append(self.cells[index - self.width])
        # RIGHT
        if ((index + 1) % self.width != 0):
            neighbours_cells.append(self.cells[index + 1])
        # BOTTOM
        if (index < len(self.cells) - self.width):
            neighbours_cells.append(self.cells[index + self.width])
        # LEFT
        if (index % self.width != 0):
            neighbours_cells.append(self.cells[index - 1])

        neighbours_cells = [cell for cell in neighbours_cells if cell.is_enabled]

        return neighbours_cells
    
    def is_next_cell_enabled(self,x,y,dir_to_move,visitable_only):
        new_posx,new_posy = self.get_relative_position(x,y,dir_to_move)
        index1 = self.to_index(new_posx,new_posy)
        index2 = self.to_index(x,y)
        if visitable_only and self.has_inter_wall(index1, index2):
            return False
        return self.cells[index1].is_enabled

    def get_allowed_directions(self,x,y,visitable_only=True):
        directions = []

        if (y > 0):
            directions.append(Direction.TOP)
        if (x < self.width - 1):
            directions.append(Direction.RIGHT)
        if (y < self.height - 1):
            directions.append(Direction.BOTTOM)
        if (x > 0):
            directions.append(Direction.LEFT)

        directions = [dir for dir in directions if self.is_next_cell_enabled(x,y,dir,visitable_only)]
        
        return directions
    
    def get_rows(self):

        rows = []
        for i in range(0, len(self.cells), self.width):
            row = [cell for cell in self.cells[i:i + self.width] if cell.is_enabled]
            if row:
                rows.append(row)

        return rows

    def remove_inter_wall(self,index1,index2):
        self.set_inter_wall_value()